### PR TITLE
[Ray Core] Make worker_process_setup_hook work with strings instead of Python functions #41096

### LINKED ISSUE #41096: [Ray Core] Make `worker_process_setup_hook` work with strings instead of Python functions
### What happened + What you expected to happen

Currently `worker_process_setup_hook` expects a python function, so it cannot be used from runtime_env that are specified via JSON (e.g. via the JobSubmission client or Ray Serve config files).

It is crucial functionality to configure logging and tracing, so we need to allow it being configured via strings.

### Versions / Dependencies

N/A

### Reproduction script

N/A

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR makes worker_process_setup_hook with work with string.

The implementation is as follow;
- We always append a special identifier if funcs are imported. If funcs are imported, we import it from the GCS and execute.
- If string is passed, we just pass the function without the special identifier. In this case, we load the module and run the setup function.

Also, it is the error message if the setup func fails
```
Failed to execute the setup hook method, ray.tests.test_runtime_env_setup_func._hooks from ``ray.init(runtime_env={'worker_process_setup_hook': ray.tests.test_runtime_env_setup_func._hooks})``. Please make sure the given module exists and is available from ray workers. For more details, see the error trace below.
Traceback (most recent call last):
  File "/Users/sangcho/work/ray/python/ray/_private/runtime_env/setup_hook.py", line 141, in load_and_execute_setup_hook_module
    setup_func = _load_class(worker_process_setup_hook_key)
  File "/Users/sangcho/work/ray/python/ray/_private/storage.py", line 491, in _load_class
    return load_class(path)
  File "/Users/sangcho/work/ray/python/ray/_private/utils.py", line 1985, in load_class
    return getattr(module, class_str)
AttributeError: module 'ray.tests.test_runtime_env_setup_func' has no attribute '_hooks'
```

## Related issue number

Closes https://github.com/ray-project/ray/issues/41096

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rickyyx — APPROVED** [review]
Does it need some documentation or if this was actually already documented with worker_process_setup_hook's usage?

File: python/ray/_private/runtime_env/setup_hook.py [hunk 1]
```diff
@@ -22,11 +25,59 @@ def get_import_export_timeout():
...
+    except Exception as e:
+        raise ray.exceptions.RuntimeEnvSetupError(
+            "Failed to export the setup function."
+        ) from e
+    env_vars = runtime_env.get("env_vars", {})
+    assert ray_constants.WORKER_PROCESS_SETUP_HOOK_ENV_VAR not in env_vars, (
+        f"The env var, {ray_constants.WORKER_PROCESS_SETUP_HOOK_ENV_VAR}, "
+        "is not permitted because it is reserved for the internal use."
+    )
+    env_vars[ray_constants.WORKER_PROCESS_SETUP_HOOK_ENV_VAR] = _decode_function_key(
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/_private/runtime_env/setup_hook.py:56]: hmm but good question. Let me check it out

[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/_private/runtime_env/setup_hook.py:56]: shouldn't be too long

**Comment by rkooo567** [comment]
> Does it need some documentation or if this was actually already documented with worker_process_setup_hook's usage?

We need more doc. I think we should actually refine this API..

**Pull request merged**